numpy/lib/twodim_base.py

```diff
@@ -32,6 +32,76 @@ def _min_int(low, high):
     return int64
 
 
+def flip(m, axis):
+    """
+    Reverse the order of elements in an array along the given axis.
+
+    All other axes are preserved, but appear in a different order than before.
+
+    Parameters
+    ----------
+    m : array_like
+        Input array.
+    axis: integer
+        Axis in array, which entries are reversed.
+        
+
+    Returns
+    -------
+    out : array_like
+        A view of `m` with the entries of axis reversed.  Since a view is
+        returned, this operation is :math:`\\mathcal O(1)`.
+
+    See Also
+    --------
+    flipud : Flip an array vertically (axis=0).
+    fliplr : Flip an array horizontally (axis=1).
+
+    Notes
+    -----
+    flip(m, 0) is equivalent to flipud(m).
+    flip(m, 1) is equivalent to fliplr(m).
+    flip(m, n) corresponds to ``A[...,::-1,...]`` with ``::-1`` at position n.
+
+    Examples
+    --------
+    >>> A = np.arange(8).reshape((2,2,2))
+    >>> A
+    array([[[0, 1],
+            [2, 3]],
+
+           [[4, 5],
+            [6, 7]]])
+
+    >>> flip(A, 0)
+    array([[[4, 5],
+            [6, 7]],
+    
+           [[0, 1],
+            [2, 3]]])
+
+    >>> flip(A, 1)
+    array([[[2, 3],
+            [0, 1]],
+    
+           [[6, 7],
+            [4, 5]]])
+
+    >>> A = np.random.randn(3,4,5)
+    >>> np.all(flip(A,2)==A[:,:,::-1,...])
+    True
+    """
+    if axis==0:
```

Comment:
Yes. 0.5 us is very, very fast.
On Fri, Feb 26, 2016 at 11:32 PM Denis Alevi <notifications@github.com>
wrote:

> In numpy/lib/twodim_base.py
> <https://github.com/numpy/numpy/pull/7346#discussion_r54328158>:
>
> > +
> > +           [[0, 1],
> > +            [2, 3]]])
> > +
> > +    >>> flip(A, 1)
> > +    array([[[2, 3],
> > +            [0, 1]],
> > +
> > +           [[6, 7],
> > +            [4, 5]]])
> > +
> > +    >>> A = np.random.randn(3,4,5)
> > +    >>> np.all(flip(A,2)==A[:,:,::-1,...])
> > +    True
> > +    """
> > +    if axis==0:
>
> The special cases increase the speed for axis = 0 and axis = 1 by a
> factor of approximately 1.5.
>
> Measuring the execution time of flip(a, 0) using timeit and for an array a
> = np.arange(100**4).reshape(100,100,100,100) I got (from 1000000 loops,
> best of 3):
>
> 0.865 us when using the special cases (returning flipud(a))
> 1.34 us when using the current implementation from abe1735
> <https://github.com/erensezener/numpy/commit/abe173528514ad71ef01d8d7c0ebf7c92abe1913>
> using @njsmith <https://github.com/njsmith> suggestion.
>
> Does that still fall into micro-optimization?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/7346/files#r54328158>.
>